teps/0041-tekton-component-versioning.md

```diff
@@ -0,0 +1,346 @@
+
+#### Disadvantages
+
+1. We should be knowing the namespace in which tekton pipelines is installed.
+2. Appropriate `Role` and `RoleBindings` are required to SSH into the pod.
+
+### 3. Creating a CRD
+
+Tekton components should create a CRD which is to provide information like
+metadata about components.
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: info.tekton.dev
+  labels:
+    app.kubernetes.io/instance: default
+    app.kubernetes.io/part-of: tekton-pipelines
+    pipeline.tekton.dev/release: "devel"
+    version: "devel"
+spec:
+  group: tekton.dev
+  versions:
+    - name: v1
+      served: true
+      storage: true
+      schema:
+        openAPIV3Schema:
+          type: object
+          properties:
+            spec:
+              type: object
+              required:
+                - version
+              properties:
+                version:
+                  type: string
+                  description: version of the component
+  scope: Cluster
+  names:
+    plural: info
+    singular: info
+    kind: Info
+    categories:
+      - tekton
+      - tekton-pipelines
+```
+
+And a CR should be added during installation which has the details
+about the version of the component or any other info required.
+
+```yaml
+apiVersion: tekton.dev/v1
+kind: Info
+metadata:
+  name: pipeline-info
+spec:
+  version: v0.18.0
+```
+
+`ClusterRole` and `ClusterRoleBinding`
+
+```yaml
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: pipelines-info-view-cr
+  labels:
+    rbac.authorization.k8s.io/aggregate-to-view: "true"
+rules:
+  - apiGroups: ["tekton.dev"]
+    resources: ["info"]
+    verbs: ["get", "list", "watch"]
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: pipelines-info-crbinding
+subjects:
+  - kind: Group
+    name: system:authenticated
+    apiGroup: rbac.authorization.k8s.io
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: pipelines-info-view
+```
+
+#### Advantages
+
+1. The CRD can be clusterwide so user need not to know the installed pipeline namespace.
+
+#### Disadvantages
+
+1. Where to store the CRD?
+
+   - Pipelines, Triggers or somewhere else?
+   - In each component then how to handle the changes?
+     - (Solution) Each component will have to maintain their own CRD so that they
+       can provide their respective information if they want to.
+
+### 4. Make a Mutating Webhook
+
+Make the mutating webhook append the version on object on creation (in an annotation).
+We can pass the version info as a flag just like we do on the controller.
+
+#### Advantages
+
+1. Easy to implement with existing codebase.
+
+#### Disadvantages
+
+1. Client will have to create a dummy Tekton resource in order to find the version.
+2. In case of `Dashboard` no dummy resource can be created.
+3. There can be a scenario where no Tekton resource is present on the cluster.
+
```

Comment:
can we add an alternative where we take the same approach as knative and make the version information available in labels on the namespace? (https://knative.dev/docs/install/check-install-version/)

Reply:
This approaches would work and wouldn't require to create a new resource (configMap). It is also really close to what we are doing now, a diff would only be the following for `tektoncd/pipeline`.

```diff
diff --git a/config/100-namespace.yaml b/config/100-namespace.yaml
index 5254a8ce..043f30b1 100644
--- a/config/100-namespace.yaml
+++ b/config/100-namespace.yaml
@@ -19,3 +19,4 @@ metadata:
   labels:
     app.kubernetes.io/instance: default
     app.kubernetes.io/part-of: tekton-pipelines
+    pipeline.tekton.dev/release: "devel"
```

The problem is that only `tektoncd/pipeline` component defines this (applies the namespace) and if other component start to define this (in a `100-namespace.yaml` file), the last installed component wins… The operator could help here, but the problem wouldn't be solved for when you install it on your own.